File: GatewayApp/app/src/main/java/com/mobilemoney/gateway/util/SMSParser.kt
package com.mobilemoney.gateway.util

import android.util.Log
import java.util.regex.Pattern

object SMSParser {
    private const val TAG = "SMSParser"

    private val FAILURE_INDICATORS = listOf(
        "failed", "unsuccessful", "declined", "insufficient", "timeout", "error",
        "not allowed", "invalid", "incorrect", "rejected", "could not"
    )

    private val PATTERNS = listOf(
        mapOf(
            "network" to "MTN",
            "success_pattern" to "(?:transaction\\s+(?:id|ref)[:\\s]+)?([A-Z0-9]{8,20})",
            "amount_pattern" to "(?:amount|amt)[:\\s]+(?:K|ZMW|MWK)?\\s*([\\d,]+\\.?\\d*)",
            "balance_pattern" to "(?:new\\s+balance|bal)[:\\s]+(?:K|ZMW|MWK)?\\s*([\\d,]+\\.?\\d*)"
        ),
        mapOf(
            "network" to "Airtel",
            "success_pattern" to "(?:transaction\\s+(?:id|ref)[:\\s]+)?([A-Z0-9]{8,20})",
            "amount_pattern" to "(?:amount|amt)[:\\s]+(?:K|ZMW|MWK)?\\s*([\\d,]+\\.?\\d*)",
            "balance_pattern" to "(?:new\\s+balance|bal)[:\\s]+(?:K|ZMW|MWK)?\\s*([\\d,]+\\.?\\d*)"
        ),
        mapOf(
            "network" to "Zamtel",
            "success_pattern" to "(?:transaction\\s+(?:id|ref)[:\\s]+)?([A-Z0-9]{8,20})",
            "amount_pattern" to "(?:amount|amt)[:\\s]+(?:K|ZMW|MWK)?\\s*([\\d,]+\\.?\\d*)",
            "balance_pattern" to "(?:new\\s+balance|bal)[:\\s]+(?:K|ZMW|MWK)?\\s*([\\d,]+\\.?\\d*)"
        )
    )

    data class ParsedSMS(
        val transactionId: String?,
        val amount: Double?,
        val newBalance: Double?,
        val reference: String?,
        val status: String,
        val rawMessage: String,
        val sender: String? = null
    )

    fun parse(message: String, sender: String? = null): ParsedSMS {
        val lower = message.lowercase()
        val isFailure = FAILURE_INDICATORS.any { lower.contains(it) }

        if (isFailure) {
            return ParsedSMS(
                transactionId = null,
                amount = null,
                newBalance = null,
                reference = null,
                status = "failed",
                rawMessage = message,
                sender = sender
            )
        }

        val transactionId = extract(message, "success_pattern")
        val amount = extractAmount(message)
        val newBalance = extractBalance(message)
        val status = if (transactionId != null || amount != null || newBalance != null) "success" else "unrecognized"

        return ParsedSMS(
            transactionId = transactionId,
            amount = amount,
            newBalance = newBalance,
            reference = transactionId,
            status = status,
            rawMessage = message,
            sender = sender
        )
    }

    fun isMnoNotification(message: String): Boolean {
        val keywords = listOf(
            "transaction", "received", "sent", "balance", "airtime", "mobile money",
            "mtn", "airtel", "zamtel", "mno", "payment", "confirmed"
        )
        val lower = message.lowercase()
        return keywords.any { lower.contains(it) }
    }

    private fun extract(message: String, key: String): String? {
        for (patternDef in PATTERNS) {
            val pattern = patternDef[key] as? String ?: continue
            val matcher = Pattern.compile(pattern, Pattern.CASE_INSENSITIVE).matcher(message)
            if (matcher.find()) {
                return matcher.group(1)?.trim()
            }
        }
        return null
    }

    private fun extractAmount(message: String): Double? {
        for (patternDef in PATTERNS) {
            val pattern = patternDef["amount_pattern"] as? String ?: continue
            val matcher = Pattern.compile(pattern, Pattern.CASE_INSENSITIVE).matcher(message)
            if (matcher.find()) {
                val raw = matcher.group(1)?.replace(",", "")
                return raw?.toDoubleOrNull()
            }
        }
        val generic = Pattern.compile("(?:K|ZMW|MWK)?\\s*([\\d,]+\\.?\\d*)", Pattern.CASE_INSENSITIVE).matcher(message)
        if (generic.find()) {
            return generic.group(1)?.replace(",", "")?.toDoubleOrNull()
        }
        return null
    }

    private fun extractBalance(message: String): Double? {
        for (patternDef in PATTERNS) {
            val pattern = patternDef["balance_pattern"] as? String ?: continue
            val matcher = Pattern.compile(pattern, Pattern.CASE_INSENSITIVE).matcher(message)
            if (matcher.find()) {
                val raw = matcher.group(1)?.replace(",", "")
                return raw?.toDoubleOrNull()
            }
        }
        return null
    }
}

File: mobile_money_gateway/utils/sms_parser.py
import re
from typing import Optional, Dict, Any
from dataclasses import dataclass


@dataclass
class ParsedSMS:
    transaction_id: Optional[str]
    amount: Optional[float]
    new_balance: Optional[float]
    reference: Optional[str]
    status: str
    raw_message: str
    sender: Optional[str] = None


class SMSParser:
    PATTERNS = [
        {
            "network": "MTN",
            "success_pattern": r"(?:Y'ello.*(?:received|sent).*ID[:\s]*|Txn\s*ID[:\s]*|ref[:\s]*)([0-9]{10,20})",
            "amount_pattern": r"(?:received|sent|amount)[:\s]*(?:K|ZMW)\s*([\d,]+\.?\d*)",
            "balance_pattern": r"(?:new\s+balance|available\s+balance)[:\s]*(?:K|ZMW)\s*([\d,]+\.?\d*)",
        },
        {
            "network": "Airtel",
            "success_pattern": r"(?:Trans\.?\s*ID[:\s]*|Txn\s*ID[:\s]*|ref[:\s]*)([A-Z0-9]{8,20})",
            "amount_pattern": r"(?:received|sent|amount)[:\s]*(?:K|ZMW)\s*([\d,]+\.?\d*)",
            "balance_pattern": r"(?:new\s+balance|available\s+balance)[:\s]*(?:K|ZMW)\s*([\d,]+\.?\d*)",
        },
        {
            "network": "Zamtel",
            "success_pattern": r"(?:Trans\.?\s*ID[:\s]*|Txn\s*ID[:\s]*|ref[:\s]*)([A-Z0-9]{8,20})",
            "amount_pattern": r"(?:received|sent|amount)[:\s]*(?:K|ZMW)\s*([\d,]+\.?\d*)",
            "balance_pattern": r"(?:new\s+balance|available\s+balance)[:\s]*(?:K|ZMW)\s*([\d,]+\.?\d*)",
        },
    ]

    FAILURE_INDICATORS = [
        "failed", "unsuccessful", "declined", "insufficient", "timeout", "error",
        "not allowed", "invalid", "incorrect", "rejected", "could not",
    ]

    @classmethod
    def parse(cls, message: str, sender: Optional[str] = None) -> ParsedSMS:
        message_lower = message.lower()
        is_failure = any(indicator in message_lower for indicator in cls.FAILURE_INDICATORS)

        if is_failure:
            return ParsedSMS(
                transaction_id=None,
                amount=None,
                new_balance=None,
                reference=None,
                status="failed",
                raw_message=message,
                sender=sender,
            )

        transaction_id = cls._extract(message, "success_pattern")
        amount = cls._extract_amount(message)
        new_balance = cls._extract_balance(message)

        status = "success" if (transaction_id or amount or new_balance) else "unrecognized"

        return ParsedSMS(
            transaction_id=transaction_id,
            amount=amount,
            new_balance=new_balance,
            reference=transaction_id,
            status=status,
            raw_message=message,
            sender=sender,
        )

    @classmethod
    def _extract(cls, message: str, pattern_key: str):
        for pattern_def in cls.PATTERNS:
            pattern = pattern_def.get(pattern_key)
            if pattern:
                match = re.search(pattern, message, re.IGNORECASE)
                if match:
                    return match.group(1).strip()
        return None

    @classmethod
    def _extract_amount(cls, message: str) -> Optional[float]:
        for pattern_def in cls.PATTERNS:
            pattern = pattern_def.get("amount_pattern")
            if pattern:
                match = re.search(pattern, message, re.IGNORECASE)
                if match:
                    raw = match.group(1).replace(",", "")
                    try:
                        return float(raw)
                    except ValueError:
                        continue
        generic_match = re.search(r"(?:K|ZMW|MWK)?\s*([\d,]+\.?\d*)", message)
        if generic_match:
            try:
                return float(generic_match.group(1).replace(",", ""))
            except ValueError:
                pass
        return None

    @classmethod
    def _extract_balance(cls, message: str) -> Optional[float]:
        for pattern_def in cls.PATTERNS:
            pattern = pattern_def.get("balance_pattern")
            if pattern:
                match = re.search(pattern, message, re.IGNORECASE)
                if match:
                    raw = match.group(1).replace(",", "")
                    try:
                        return float(raw)
                    except ValueError:
                        continue
        return None

    @classmethod
    def is_mno_notification(cls, message: str) -> bool:
        mno_keywords = [
            "transaction", "received", "sent", "balance", "airtime", "mobile money",
            "mtn", "airtel", "zamtel", "mno", "payment", "confirmed",
        ]
        message_lower = message.lower()
        return any(keyword in message_lower for keyword in mno_keywords)
